### PR TITLE
[TableGen][SubtargetEmitter] Early exit from loop in FindWriteResources and FindReadAdvance

### PR DESCRIPTION
**Pull request by michaelmaitland**
This gives us a 30% speed improvement in our downstream.

### CODE REVIEW

File: llvm/utils/TableGen/SubtargetEmitter.cpp [hunk 1]
```diff
@@ -910,6 +910,7 @@ SubtargetEmitter::FindWriteResources(const CodeGenSchedRW &SchedWrite,
                                           ProcModel.ModelName);
       }
       ResDef = WR;
+      break;
```
Comment (topperc) [inline llvm/utils/TableGen/SubtargetEmitter.cpp:913]: Doesn't this make the PrintFatalError dead code?

Comment (michaelmaitland) [reply llvm/utils/TableGen/SubtargetEmitter.cpp:913]: I messed up. Need to take a second look at this.

Comment (michaelmaitland) [reply llvm/utils/TableGen/SubtargetEmitter.cpp]: I have addressed this problem

File: llvm/utils/TableGen/SubtargetEmitter.cpp [hunk 2]
```diff
@@ -902,8 +902,13 @@ SubtargetEmitter::FindWriteResources(const CodeGenSchedRW &SchedWrite,
...
       continue;
-    if (AliasDef == WR->getValueAsDef("WriteType") ||
-        SchedWrite.TheDef == WR->getValueAsDef("WriteType")) {
+    // If there is no AliasDef and we find a match, we can early exit since
+    // there is no need to verify whether there are resources defined for both
+    // SchedWrite and its alias.
+    if (!AliasDef && SchedWrite.TheDef == WR->getValueAsDef("WriteType")) {
+      ResDef = WR;
+      break;
+    } else if (AliasDef == WR->getValueAsDef("WriteType")) {
```
[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/utils/TableGen/SubtargetEmitter.cpp]: updated

File: llvm/utils/TableGen/SubtargetEmitter.cpp [hunk 3]
```diff
@@ -902,8 +902,14 @@ SubtargetEmitter::FindWriteResources(const CodeGenSchedRW &SchedWrite,
   for (Record *WR : ProcModel.WriteResDefs) {
     if (!WR->isSubClassOf("WriteRes"))
       continue;
-    if (AliasDef == WR->getValueAsDef("WriteType") ||
-        SchedWrite.TheDef == WR->getValueAsDef("WriteType")) {
+    // If there is no AliasDef and we find a match, we can early exit since
+    // there is no need to verify whether there are resources defined for both
+    // SchedWrite and its alias.
+    if (!AliasDef && SchedWrite.TheDef == WR->getValueAsDef("WriteType")) {
```
Comment (topperc) [inline llvm/utils/TableGen/SubtargetEmitter.cpp:908]: While you're here can we cache `WR->getValueAsDef("WriteType")` in a local variable. We call it twice.

File: llvm/utils/TableGen/SubtargetEmitter.cpp [hunk 4]
```diff
@@ -956,8 +962,14 @@ Record *SubtargetEmitter::FindReadAdvance(const CodeGenSchedRW &SchedRead,
   for (Record *RA : ProcModel.ReadAdvanceDefs) {
     if (!RA->isSubClassOf("ReadAdvance"))
       continue;
-    if (AliasDef == RA->getValueAsDef("ReadType") ||
-        SchedRead.TheDef == RA->getValueAsDef("ReadType")) {
+    // If there is no AliasDef and we find a match, we can early exit since
+    // there is no need to verify whether there are resources defined for both
+    // SchedWrite and its alias.
+    if (!AliasDef && AliasDef == RA->getValueAsDef("ReadType")) {
```
Comment (topperc) [inline llvm/utils/TableGen/SubtargetEmitter.cpp]: Cache RA->getValueAsDef("ReadType")

File: llvm/utils/TableGen/SubtargetEmitter.cpp (hunk 3 above)
Comment (michaelmaitland) [reply llvm/utils/TableGen/SubtargetEmitter.cpp]: you got it. done

File: llvm/utils/TableGen/SubtargetEmitter.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (michaelmaitland) [reply llvm/utils/TableGen/SubtargetEmitter.cpp]: yes. updated.

File: llvm/utils/TableGen/SubtargetEmitter.cpp [hunk 5]
```diff
@@ -902,8 +902,15 @@ SubtargetEmitter::FindWriteResources(const CodeGenSchedRW &SchedWrite,
   for (Record *WR : ProcModel.WriteResDefs) {
     if (!WR->isSubClassOf("WriteRes"))
       continue;
-    if (AliasDef == WR->getValueAsDef("WriteType") ||
-        SchedWrite.TheDef == WR->getValueAsDef("WriteType")) {
+    // If there is no AliasDef and we find a match, we can early exit since
+    // there is no need to verify whether there are resources defined for both
+    // SchedWrite and its alias.
+    Record *WRDef = WR->getValueAsDef("WriteType");
+    if (!AliasDef && SchedWrite.TheDef == WRDef) {
```
Comment (topperc) [inline llvm/utils/TableGen/SubtargetEmitter.cpp:909]: What if AliasDef is non-null. We won't update ResDef if SchedWrite.TheDef == WRDef.

```suggestion
Record *WRDef = WR->getValueAsDef("WriteType");
if (AliasDef == WRDef || SchedWrite.TheDef == WRDef) {
  if (ResDef)
    PrintFatalError...
   ResDef = WR;
   // If there is no AliasDef and we find a match, we can early exit...
   if (!AliasDef)
     break;
```

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**